private/buf/cmd/buf/command/alpha/repo/reposync/reposync.go

```diff
@@ -0,0 +1,346 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reposync
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/bufbuild/buf/private/buf/bufcli"
+	"github.com/bufbuild/buf/private/buf/bufsync"
+	"github.com/bufbuild/buf/private/bufpkg/bufanalysis"
+	"github.com/bufbuild/buf/private/bufpkg/bufmanifest"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmoduleref"
+	"github.com/bufbuild/buf/private/gen/proto/connect/buf/alpha/registry/v1alpha1/registryv1alpha1connect"
+	registryv1alpha1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/registry/v1alpha1"
+	"github.com/bufbuild/buf/private/pkg/app/appcmd"
+	"github.com/bufbuild/buf/private/pkg/app/appflag"
+	"github.com/bufbuild/buf/private/pkg/command"
+	"github.com/bufbuild/buf/private/pkg/connectclient"
+	"github.com/bufbuild/buf/private/pkg/git"
+	"github.com/bufbuild/buf/private/pkg/manifest"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"github.com/bufbuild/buf/private/pkg/storage/storagegit"
+	"github.com/bufbuild/buf/private/pkg/stringutil"
+	"github.com/bufbuild/connect-go"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	errorFormatFlagName      = "error-format"
+	moduleFlagName           = "module"
+	createFlagName           = "create"
+	createVisibilityFlagName = "create-visibility"
+)
+
+// NewCommand returns a new Command.
+func NewCommand(
+	name string,
+	builder appflag.Builder,
+) *appcmd.Command {
+	flags := newFlags()
+	return &appcmd.Command{
+		Use:   name,
+		Short: "Sync a Git repository to a registry",
+		Long: "Sync a Git repository's commits to a registry in topological order. " +
+			"Only commits belonging to the 'origin' remote are processed, which means that " +
+			"commits must be pushed to a remote. " +
+			"Only modules specified via '--module' are synced.",
+		Args: cobra.NoArgs,
+		Run: builder.NewRunFunc(
+			func(ctx context.Context, container appflag.Container) error {
+				return run(ctx, container, flags)
+			},
+			bufcli.NewErrorInterceptor(),
+		),
+		BindFlags: flags.Bind,
+	}
+}
+
+type flags struct {
+	ErrorFormat      string
+	Modules          []string
+	Create           bool
+	CreateVisibility string
+}
+
+func newFlags() *flags {
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	flagSet.StringVar(
+		&f.ErrorFormat,
+		errorFormatFlagName,
+		"text",
+		fmt.Sprintf(
+			"The format for build errors printed to stderr. Must be one of %s",
+			stringutil.SliceToString(bufanalysis.AllFormatStrings),
+		),
+	)
+	flagSet.StringArrayVar(
+		&f.Modules,
+		moduleFlagName,
+		nil,
+		"The module(s) to sync to the BSR; you can provide a module override in the format "+
```

Comment:
Is this module dirs or module identities? I mean:

```
--module some/repo/dir:buf.build/user/overridemodule

or

--module buf.build/user/somelocalbufyaml:buf.build/user/overridemodule
```

Reply:
There's an interesting challenge here that we've thought at some point when syncing third-party modules. What to do, or how to handle commits in which the module dir changes... or even worse, when it's swapped with other module dir.

Reply:
Module dir. This in fact lets us handle the first case you've identified, but if the dir is swapped, we're out of luck.

### Example 1 (handled)

```
commit a: 
- module in dir `proto` points to `buf.build/bufbuild/bufapis`
commit b: 
- module in dir `proto` points to `buf.build/bufbuild/buf`
commit c: 
- module in dir `protos` points to `buf.build/bufbuild/buf`
```

To sync this "properly", the user would want to run
```
buf sync --module proto:buf.build/bufbuild/buf --module protos:buf.build/bufbuild/buf
```
 and we'll "follow" the module how they intended. "follow" because it's not strictly a follow, but the end result is the same.

### Example 2 (handled)

```
commit a: 
- module in dir `proto` points to `buf.build/bufbuild/buf`
commit b: 
- module in dir `proto/buf` points to `buf.build/bufbuild/buf`
- module in dir `proto/audit` points to `buf.build/bufbuild/audit`
```
To sync this "properly", the user would want to run
```
buf sync --module proto:buf.build/bufbuild/buf --module proto/buf:buf.build/bufbuild/buf --module proto/audit:buf.build/bufbuild/audit
```

### Example 3 (unhandled)

```
commit a: 
- module in dir `proto/buf` points to `buf.build/bufbuild/buf`
- module in dir `proto/audit` points to `buf.build/bufbuild/audit`
commit b: switch identity 
- module in dir `proto/audit` points to `buf.build/bufbuild/buf`
- module in dir `proto/buf` points to `buf.build/bufbuild/audit`
```

We cannot follow this on our flags alone, we would have to look at the commit and see if the `buf.yaml` moved, and try to follow that. Following the `buf.yaml` might be more resilient long term, but if the module identity is changed at the same time, this can get really tricky. Eventually if this becomes a thing, we may have to introduce a `buf.sync.yaml` for cases like this.

Related: we cannot sync on module identity alone, as people may have test files with the same identity, we don't want to sync those. We could exclude certain dirs from being walked, but that looks error prone. I suspect we'll see some of this come up during the alpha and we can re-evaluate then.

Reply:
Thanks for the detailed explanation, yeah those edge cases will be tricky if/when they present themselves. Hopefully users don't switch their proto dirs too often with _another_ proto dir.

A `buf.sync.yaml` might sound like the perfect solution, but I think the whole reason of why `buf sync` is needed is to recover previous history into the BSR, which is most likely not buf-aware, and won't have such file in the history.

Reply:
Yeah we'd read `buf.sync.yaml` from the `origin/head`, but we can come to that if needed.